frontend/src/metabase-lib/v1/parameters/utils/targets.ts

```diff
@@ -147,29 +155,89 @@ export function buildTextTagTarget(tagName: string): ParameterTextTarget {
+export function getParameterColumns(
+  question: Question,
+  parameter?: Parameter,
+): {
+  query: Lib.Query;
+  columns: {
+    stageIndex: number;
+    column: Lib.ColumnMetadata;
+    group: Lib.ColumnGroup;
+  }[];
+} {
   // treat the dataset/model question like it is already composed so that we can apply
   // dataset/model-specific metadata to the underlying dimension options
   const query =
     question.type() !== "question"
       ? question.composeQuestionAdhoc().query()
       : question.query();
-  const stageIndex = -1;
-  const availableColumns =
-    parameter && isTemporalUnitParameter(parameter)
-      ? Lib.breakouts(query, stageIndex).map(breakout =>
-          Lib.breakoutColumn(query, stageIndex, breakout),
-        )
-      : Lib.filterableColumns(query, stageIndex);
-  const filteredColumns = parameter
-    ? availableColumns.filter(
-        columnFilterForParameter(query, stageIndex, parameter),
+
+  if (parameter && isTemporalUnitParameter(parameter)) {
+    const availableColumns = getTemporalColumns(query);
+    const columns = availableColumns.filter(({ column, stageIndex }) => {
+      return columnFilterForParameter(query, stageIndex, parameter)(column);
+    });
+
+    return { query, columns };
+  }
+
+  const nextQuery = appendStageIfAggregated(query);
+  const availableColumns = getFilterableColumns(nextQuery);
+  const columns = parameter
+    ? availableColumns.filter(({ column, stageIndex }) =>
+        columnFilterForParameter(nextQuery, stageIndex, parameter)(column),
       )
     : availableColumns;
 
-  return {
-    query,
+  return { query: nextQuery, columns };
+}
+
+function getTemporalColumns(query: Lib.Query) {
+  const stageIndex = -1;
+  const columns = Lib.breakouts(query, stageIndex).map(breakout => {
+    return Lib.breakoutColumn(query, stageIndex, breakout);
+  });
+  const [group] = Lib.groupColumns(columns);
+
+  return columns.map(column => ({
     stageIndex,
-    columns: filteredColumns,
-  };
+    column,
+    group,
+  }));
+}
+
+function getFilterableColumns(query: Lib.Query) {
+  const stageIndexes = getFilterStageIndexes(query);
+
+  return stageIndexes.flatMap(stageIndex => {
+    const columns = Lib.filterableColumns(query, stageIndex);
+    const groups = Lib.groupColumns(columns);
+
+    return groups.flatMap(group => {
+      const columns = Lib.getColumnsFromColumnGroup(group);
+
+      return columns.map(column => ({
+        stageIndex,
+        column,
+        group,
+      }));
+    });
+  });
+}
+
+function appendStageIfAggregated(query: Lib.Query) {
+  const aggregations = Lib.aggregations(query, -1);
+  const breakouts = Lib.breakouts(query, -1);
+
+  return aggregations.length > 0 && breakouts.length > 0
+    ? Lib.appendStage(query)
+    : query;
+}
+
+/**
+ * Returns indexes of stages from which columns are exposed for filtering
+ */
+function getFilterStageIndexes(query: Lib.Query): number[] {
+  return Lib.stageCount(query) > 1 ? [-2, -1] : [-1];
```

Comment:
I might be understanding this PR, but why not allow columns from stages that are further back than 1 or 2.

What if there are three stages or more?

Reply:
> Off topic and I might be understanding this PR, but why not allow columns from stages that are further back than 1 or 2.

We actually discussed this during kick-off - there's no reason not to do this but it'll be a separate project (see "Separate projects" in [product doc](https://www.notion.so/metabase/Allow-to-use-both-pre-and-post-last-aggregation-columns-in-dashboard-filters-b32a6bb5cebc4723b8b9b81870217a79))

> Is it just to stay consistent with the filters on the dashboard?

To stay consistent with the filters in filter modal in chill mode.

> I tested that locally in the PR and it does not seem to work.

It's not supposed to work yet, this PR is only about showing these columns in UI.
I updated PR description.
If something does not match the PR description - please let me know with repro steps :pray: